// pareto-core-internals/dist/imp/public/wrapRawArray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapRawArray = wrapRawArray;
const createCounter_1 = require("../private/createCounter");
const wrapAsyncValueImp_1 = require("./wrapAsyncValueImp");
/**
 * this is an implementation, not public by design
 * If you feel the need to rename this class, don't rename it to 'Array',
 * it will break the 'instanceOf Array' test
 */
class ArrayClass {
    constructor(data) {
        this.data = data;
    }
    map($v) {
        return wrapRawArray(this.data.map((entry) => {
            return $v(entry);
        }));
    }
    asyncMap($v) {
        // const elements = source.map($v)
        // let _isGuaranteedToReturnAResult = true
        // source.forEach(($) => {
        //     if ($)
        // })
        function array(array, $v) {
            const mapped = array.map($v);
            return (0, wrapAsyncValueImp_1.wrapAsyncValueImp)(($c) => {
                const temp = [];
                (0, createCounter_1.createCounter)((counter) => {
                    mapped.forEach((v) => {
                        counter.increment();
                        v.__execute((v) => {
                            temp.push(v);
                            counter.decrement();
                        });
                    });
                }, () => {
                    $c(wrapRawArray(temp));
                });
            });
        }
        return array(this.data, $v);
    }
    /////////
    __forEach($i) {
        this.data.forEach(($) => {
            $i($);
        });
    }
    __getLength() {
        return this.data.length;
    }
    __getElementAt(index) {
        return this.data[index];
    }
}
/**
 * returns a Pareto array
 * why is this not the constructor? to call a constructor, you have to use the keyword 'new'. Pareto doesn't use the concept of a class so that keyword should be avoided

 * @param source An array literal
 * @returns
 */
function wrapRawArray(source) {
    const data = source.slice(); //create a copy
    if (!(data instanceof Array)) {
        throw new Error("invalid input in 'createArray'");
    }
    return new ArrayClass(source);
}
